gk/main.c

```diff
@@ -542,6 +612,103 @@ setup_gk_instance(unsigned int lcore_id, struct gk_config *gk_conf)
 	return ret;
 }
 
+static struct flow_entry *
+find_flow_entry_candidate(struct gk_instance *instance,
+	uint32_t bidx, uint64_t request_timeout_cycles,
+	enum gk_flow_state state_to_add)
+{
+	int32_t index;
+	uint32_t next = 0;
+	const struct ip_flow *key;
+	struct flow_entry *last_fe = NULL;
+	void *data;
+	uint64_t now = rte_rdtsc();
+
+	index = rte_hash_bucket_iterate(instance->ip_flow_hash_table,
+		(void *)&key, &data, &bidx, &next);
+	while (index >= 0) {
+		struct flow_entry *fe = &instance->ip_flow_entry_table[index];
+
+		/* Expired flow entry. */
+		if (is_flow_expired(fe, now, request_timeout_cycles))
+			return fe;
+
+		/*
+		 * Only flow entries with state GK_REQUEST
+		 * will be possibly repaced, others have a higher priority.
+		 */
+		if (fe->state == GK_REQUEST) {
+			uint8_t priority = priority_from_delta_time(now,
+				fe->u.request.last_packet_seen_at);
+			/*
+			 * Do not favor request entries that are not doubling
+			 * its priority when a Gatekeeper server is overloaded.
+			 * We use +2 instead of +1 in the test below to account
+			 * for random delays in the network.
+			 */
+			if (priority > fe->u.request.last_priority + 2)
+				return fe;
+
+			if (state_to_add != GK_REQUEST && (last_fe == NULL ||
+					last_fe->u.request.last_packet_seen_at >
+					fe->u.request.last_packet_seen_at))
+				last_fe = fe;
+		}
+
+		index = rte_hash_bucket_iterate(instance->ip_flow_hash_table,
+			(void *)&key, &data, &bidx, &next);
+	}
+
+	return last_fe;
+}
+
+static int
+drop_flow_entry_heuristically(struct gk_instance *instance,
+	hash_sig_t sig, uint64_t request_timeout_cycles,
+	enum gk_flow_state state_to_add)
+{
+	uint32_t primary_bidx = rte_hash_get_primary_bucket(
+		instance->ip_flow_hash_table, sig);
+	struct flow_entry *fe = find_flow_entry_candidate(
+		instance, primary_bidx, request_timeout_cycles, state_to_add);
+	if (fe == NULL)
+		return -ENOSPC;
+
+	return gk_del_flow_entry_from_hash(instance->ip_flow_hash_table, fe);
+}
+
+/*
+ * We heuristically drop entries to alleviate memory pressure
+ * when the table is full.
+ */
+static int
+gk_hash_add_flow_entry(struct gk_instance *instance,
+	struct ip_flow *flow, unsigned int request_timeout_cycles,
+	uint32_t rss_hash_val, enum gk_flow_state state_to_add)
+{
+	int ret;
+
+hash_add_flow:
+	ret = rte_hash_add_key_with_hash(
+		instance->ip_flow_hash_table, flow, rss_hash_val);
+	if (ret == -EINVAL) {
```

Comment:
Instead of testing for `-EINVAL`, test for `ret < 0` and include `ret` in the error message. This makes the code generic and we can even leave this `if` statement after the loop and before the final return.